pkgs/development/interpreters/php/service.nix

```
@@ -0,0 +1,182 @@
+          '';
+        };
+
+        "pm.max_children" = lib.mkOption {
+          type = lib.types.int;
+          description = ''
+            The number of child processes to be created when `pm` is set to `static` and the maximum
+            number of child processes to be created when `pm` is set to `dynamic`.
+
+            This option sets the limit on the number of simultaneous requests that will be served.
+          '';
+        };
+
+        user = lib.mkOption {
+          type = lib.types.str;
+          description = ''
+            Unix user of FPM processes.
+          '';
+        };
+      };
+    };
+in
+{
+  _class = "service";
+
+  options.php-fpm = {
+    package = lib.mkPackageOption pkgs "php" {
+      example = ''
+        php.buildEnv {
+          extensions =
+            { all, ... }:
+            with all;
+            [
+              imagick
+              opcache
+            ];
+          extraConfig = "memory_limit=256M";
+        }
+      '';
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType =
+          with lib.types;
+          attrsOf (oneOf [
+            str
+            int
+            bool
+            (submodule poolOpts)
+          ]);
+        options = {
+          log_level = lib.mkOption {
+            type = lib.types.enum [
+              "alert"
+              "error"
+              "warning"
+              "notice"
+              "debug"
+            ];
+            default = "notice";
+            description = ''
+              Error log level.
+            '';
+          };
+        };
+      };
+      default = { };
+      example = lib.literalExpression ''
+        {
+          log_level = "debug";
+          log_limit = 2048;
+
+          mypool = {
+            "user" = "php";
+            "group" = "php";
+            "listen.owner" = "caddy";
+            "listen.group" = "caddy";
+            "pm" = "dynamic";
+            "pm.max_children" = 75;
+            "pm.start_servers" = 10;
+            "pm.min_spare_servers" = 5;
+            "pm.max_spare_servers" = 20;
+            "pm.max_requests" = 500;
+          }
+        }
+      '';
+      description = ''
+        PHP FPM configuration. Refer to [upstream documentation](https://www.php.net/manual/en/install.fpm.configuration.php) for details on supported values.
+      '';
+    };
+  };
+
+  config = {
+    php-fpm.settings = {
+      error_log = "syslog";
+      daemonize = false;
+    };
+
+    process.argv = [
+      "${cfg.package}/bin/php-fpm"
+      "-y"
+      configFile
+    ];
+  }
+  // lib.optionalAttrs (options ? systemd) {
+
+    systemd.service = {
+      after = [ "network.target" ];
+      documentation = [ "man:php-fpm(8)" ];
+
+      serviceConfig = {
+        Type = "notify";
+        ExecReload = "${pkgs.coreutils}/bin/kill -USR2 $MAINPID";
```

```suggestion
        ExecReload = toString [
          (lib.getExe' pkgs.coreutils "kill")
          "-USR2"
          "$MAINPID"
        ];
```

<del>Sorry, I have a commandline tokenization kink.